### PR TITLE
5.1 rethreading

### PR DESCRIPTION
**Pull request by kkekkonen**
#### Summary
Adds a new feature: rethreading. Posts not included in threads can be moved to existing threads, or a new thread can be created with another post not in threads. Rethreading can be done by the posts creator (and any system admin or team admin if he/she has permissions to edit the post).

#### Ticket Link
None. 

There is the Mattermost forum post:
https://mattermost.uservoice.com/forums/306457-general/suggestions/34136164-add-ability-to-move-mistaken-thread-replies-in-a-c

and mattermost feature channel
https://pre-release.mattermost.com/core/pl/13zruhcrfidtxqi9s6m8tu54fe

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] Ran `make check-style` to check for style errors (required for all pull requests)
- [x] Ran `make test` to ensure unit and component tests passed
- [ ] Added or updated unit tests (required for all new features)
- [x] Has server changes (please link) https://github.com/mattermost/mattermost-server/pull/9165
- [x] Has redux changes (please link) https://github.com/mattermost/mattermost-redux/pull/581
- [x] Has UI changes
- [ ] Includes text changes and localization file ([.../i18n/en.json](https://github.com/mattermost/mattermost-webapp/blob/master/i18n/en.json)) updates
- [ ] Touches critical sections of the codebase (auth, posting, etc.)

Note: package.json uses mattermost-redux from kkekkonen/mattermost-redux. The conflicts can be resolved after rethreading is added to mattermost-redux

### CODE REVIEW

File: components/post_view/post_body/post_body.jsx [hunk 1]
```diff
@@ -92,6 +92,11 @@ export default class PostBody extends React.PureComponent {
          * Set not to allow edits on post
          */
         isReadOnly: PropTypes.bool,
+
+        /**
+         * Set to mark the post as selected for rethreding
```
Comment (ThiefMaster) [inline components/post_view/post_body/post_body.jsx]: typo -> rethreading

File: package.json [hunk 2]
```diff
@@ -26,7 +26,7 @@
     "localforage": "1.7.2",
     "localforage-observable": "1.4.0",
     "marked": "github:mattermost/marked#ed33baecd7d7fa97d479ba22dde9d226b083d67d",
-    "mattermost-redux": "github:mattermost/mattermost-redux#c3a7215403d8607a21a30e11f6066308c448ce75",
+    "mattermost-redux": "github:kkekkonen/mattermost-redux#mattermost-redux-rethreading",
```
Comment (ThiefMaster) [inline package.json]: :warning: probably shouldn't be part of the PR

File: utils/post_utils.jsx [hunk 3]
```diff
@@ -93,6 +93,38 @@ export function canEditPost(post, editDisableAction) {
...
+
+    let canRethread = false;
+    let canRethreadOthers = false;
+    const license = getLicense(store.getState());
+    const config = getConfig(store.getState());
+    const channel = getChannel(store.getState(), post.channel_id);
+
+    const isOwner = isPostOwner(post);
+    canRethread = haveIChannelPermission(store.getState(), {channel: post.channel_id, team: channel && channel.team_id, permission: Permissions.EDIT_POST});
+    canRethreadOthers = haveIChannelPermission(store.getState(), {channel: post.channel_id, team: channel && channel.team_id, permission: Permissions.EDIT_OTHERS_POSTS});
```
Comment (ThiefMaster) [inline utils/post_utils.jsx]: does it really make sense to use the "edit other posts" permission for this? as someone running a mattermost server I would probably NOT allow (m)any people  to edit others' posts, but allowing rethreading is much less invasive and something i'd probably let every team admin do, or even a channel admin.

to me it sounds like having a separate permission for this would be better UX, especially for the "others" version

File: utils/post_utils.jsx [hunk 4]
```diff
@@ -93,6 +93,38 @@ export function canEditPost(post, editDisableAction) {
...
+    if (!isOwner && !canRethreadOthers) {
+        canRethread = false;
+    }
+
+    if (canRethread && license.IsLicensed === 'true') {
```
Comment (ThiefMaster) [inline utils/post_utils.jsx]: do edit time limits usually apply for editing posts of others? (if yes, isn't that a bug? i'd expect an admin who can edit posts of others to do that regardless of how old the post is)

Comment (kkekkonen) [reply utils/post_utils.jsx]: The time limit is a bit useless feature for rethreading. Do you think it should be removed entirely?

File: package.json (hunk 2 above)
Comment (kkekkonen) [reply package.json]: I guess I should wait for the mm-redux PR to be merged first

File: utils/post_utils.jsx (hunk 3 above)
Comment (kkekkonen) [reply utils/post_utils.jsx]: This needs more discussion in my opinion. The current solution "cuts corners" by using existing edit_post permission,  but I'm not entirely sure how to add new permissions and integrate them to the permission system/scheme etc.

**Comment by esethna** [comment]
Thanks @kkekkonen! If you can help get the build passing we can get a test server going and look a bit more into the UX

**Comment by kkekkonen** [comment]
The build seems to be passing now that the PR was rebased

**Comment by kkekkonen** [comment]
Made some changes to the rethreading-webapp: rethreading is now done via dragging, and posts can also be moved from thread to thread

**Comment by kkekkonen** [comment]
Now includes a experimental flag in System Console/Developer settings which can be used to enable and disable rethreading

**Comment by hmhealey** [comment]
FYI, something appears to have gone wrong when updating the branch for this PR since it seems to include copies of a bunch of unrelated commits

**Comment by kkekkonen** [comment]
I dont know why the unrelated comments were there, but I got them removed.

**Comment by cpanato** [comment]
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

**Comment by cpanato** [comment]
Spintmint test server creation failed. See the logs for more information.

**Comment by esethna** [comment]
Apologies for the delay here, let's get redux and server merged so we can test and review this PR

**Comment by jasonblais** [comment]
After discussing with @kkekkonen, I will close this and the related pull requests, but keep the branches open so work can be easily continued later on.

The work halted here as a result of our process failures. In particular,
 - there was no owner for these PRs, leading to slow response times from the Mattermost team
 - there was no ticket, nor no design, prior to submitting the PRs, leading to 
   - Mattermost team giving confusing feedback (too high-level)
   - uncertainty and inconsistency about the direction of the feature
   - conflicting reviewer feedback, in some cases leading to contributor reverting his changes back

We've taken notes and started discussions on how to address the above - we'll start by having a JIRA ticket for every contributor pull request that details the expected behaviour and design, and have an owner for each.

**Comment by enzolutions** [comment]
@jasonblais @kkekkonen This issue was closed in behalf another issue ? or maybe because was solved in a PR?

**Comment by jasonblais** [comment]
Hi @enzolutions - we closed the PR due to reasons outlined here: https://github.com/mattermost/mattermost-webapp/pull/1481#issuecomment-447855385
